plugins/newsdownloader.koplugin/feed_source.lua

```diff
@@ -0,0 +1,403 @@
+        local unsupported_urls = ""
+        for key, value in pairs(unsupported_feeds_urls) do
+            -- Create the error message.
+            --            unsupported_urls = unsupported_urls .. " " .. value[1] .. " " .. value[2]
+            unsupported_urls = value[1] .. "\n\n"
+            -- Not sure what this does.
+            if key ~= #unsupported_feeds_urls then
+                unsupported_urls = BD.url(unsupported_urls) .. ", "
+            end
+        end
+        error_callback(
+            T(_([[Could not initialize some feeds\n\n %1 \n\nPlease review your feed configuration file.]]),
+              unsupported_urls)
+        )
+    end
+
+    return initialized_feeds
+end
+
+-- This function contacts the feed website and attempts to get
+-- the RSS/Atom document with a list of the latest items.
+function FeedSource:fetchDocumentByUrl(url)
+    local document
+    -- Get the XML document representing the feed
+    local ok, response = pcall(function()
+            local success, content = NewsHelpers:getUrlContent(url)
+            if (success) then
+                return content
+            else
+                error("Failed to download content for url:", url)
+            end
+    end)
+    -- Check to see if a response is available to deserialize.
+    if ok then
+        -- Deserialize the XML document into something Lua can use
+        document = NewsHelpers:deserializeXMLString(response)
+    end
+    -- Return the document or any errors that may have occured
+    if ok or document then
+        return document
+    else
+        if not ok then
+            error("(Reason: Failed to download feed document)")
+        else
+            error("(Reason: Error during feed document deserialization)")
+        end
+    end
+end
+
+-- Supply this method with the XML document returned by the feed,
+-- and it will initialized the document by extracting the feed title,
+-- feed items, and items count.
+function FeedSource:initializeDocument(document)
+    local feed_title
+    local feed_items
+    local total_items
+
+    local ok = pcall(function()
+            return self:getFeedType(
+                document,
+                function()
+                    -- RSS callback
+                    feed_title = util.htmlEntitiesToUtf8(document.rss.channel.title)
+                    feed_items = document.rss.channel.item
+                    total_items = #document.rss.channel.item
+                end,
+                function()
+                    -- Atom callback
+                    feed_title = FeedSource:getFeedTitle(document.feed.title)
+                    feed_items = document.feed.entry
+                    total_items = #document.feed.entry
+                end
+            )
+    end)
+
+    if ok then
+        document.title = feed_title
+        document.items = feed_items
+        document.total_items = total_items
+        return document
+    else
+        error("Could not initialize feed document")
```

Comment:
Standard question, this is behind a pcall?

Reply:
Yes, originating here: 

https://github.com/koreader/koreader/pull/8263/files#diff-52bb22da5ec6daa7e58ab1c7774c871823ef8f2f1a35f348c424ffab703bedb7R36-R40